### ISSUE TITLE
Transpose methods?

### ISSUE THREAD

**Issue by clarfonthey** [rust-lang/portable-simd#306]
One thing I've been particularly interested in is a (relatively) simple SIMD matrix implementation. There are _loads_ of implementations out there for stuff like matrix multiplication, and I know that this is a gigantic rabbit hole, but I feel like there are some common operations that stand out from the crowd as having one "correct" solution, which are pretty easy to mess up.

For example, looking at the [matrix inversion](https://github.com/rust-lang/portable-simd/blob/2c5ebfb6a26d384bc21db6796095890c1f13f19c/crates/core_simd/examples/matrix_inversion.rs) example, there's a rather cryptic implementation of a 4x4 matrix transpose here:

```rust
const SHUFFLE01: [Which; 4] = [First(0), First(1), Second(0), Second(1)];
const SHUFFLE02: [Which; 4] = [First(0), First(2), Second(0), Second(2)];
const SHUFFLE13: [Which; 4] = [First(1), First(3), Second(1), Second(3)];
const SHUFFLE23: [Which; 4] = [First(2), First(3), Second(2), Second(3)];

let tmp = simd_swizzle!(m_0, m_1, SHUFFLE01);
let row1 = simd_swizzle!(m_2, m_3, SHUFFLE01);

let row0 = simd_swizzle!(tmp, row1, SHUFFLE02);
let row1 = simd_swizzle!(row1, tmp, SHUFFLE13);

let tmp = simd_swizzle!(m_0, m_1, SHUFFLE23);
let row3 = simd_swizzle!(m_2, m_3, SHUFFLE23);
let row2 = simd_swizzle!(tmp, row3, SHUFFLE02);
let row3 = simd_swizzle!(row3, tmp, SHUFFLE13);
```

It feels like it would be helpful to offer some easy methods that people can use for "transposing" a set of SIMD vectors which represent a matrix, since doing so is required for all sorts of operations. Essentially, given n SIMD m-vectors, return m SIMD n-vectors, assuming that n and m are both valid lane sizes. Probably the most necessary combinations are 2x2 and 4x4, but assuming that we write the code right, we should be able to make this work with any combination of valid lane sizes.

Maybe it might even be convenient to have a dedicated SIMD type that represents such a matrix (with valid lane sizes on both dimensions), although that may be more than you'd want for the core SIMD module.

I could also be overgeneralising something here that would be better suited for a library, but since we're already in untreaded territory (I don't know of anything similar to `std::simd` in any other language), it's worth bringing up.

**Comment by thomcc** [comment]
A transpose function is on the boundary, but in general I'd prefer this kind of thing to be in a library.

I definitely think a matrix type doesn't belong here -- often that's not the best way to do simd anyway (although it would depend on the situation) -- concretely, you want to operate in a wider fashion, since you often have many matrices you need to work on, so you'd do 4x or 8x multiplications at once, not a single one.

**Comment by calebzulawski** [comment]
Perhaps this example is not the best, since it's just a translation of some older SIMD examples.  The most straightforward way to write a transpose would probably involve generating the const indices, instead of hard-coding them.

**Comment by clarfonthey** [comment]
Oh yeah, I should add that I don't necessarily mean that we should take the examples as the source of Ideal SIMD Performance, but mostly used it as an existing example of (what I feel) most SIMD transpose operations would look like when implemented, as a justification or why it might be reasonable to include.

Teaching how to use SIMD in a world where people are portable SIMD doesn't really exist is its own can of worms that's better discussed elsewhere.

**Comment by HadrienG2** [comment]
This SIMD transpose operation also gets my vote because...

- It is quite hard to implement correctly and efficiently, but once it's done, it's not that much code to maintain.
- It is useful in a wide range of classic SIMD computation use cases including SoA/AoS conversions, stencils... and of course, linear algebra.
- People won't want to add a dependency just for this operation, so unless people have more ideas for a more full-featured "SIMD algorithms" crate, it makes sense to make it part of the common std vocabulary.

**Comment by clarfonthey** [comment]
I disagree with the notion that people won't add dependencies for small mathematical operations, but I personally feel great apprehension in pulling in a gigantic linear algebra library for the same reason.

That said, if we were to implement some kind of matrix transpose, I doubt it would be on some sort of proper matrix type and instead just on an array of SIMD vectors.

**Comment by HadrienG2** [comment]
> I disagree with the notion that people won't add dependencies for small mathematical operations, but I personally feel great apprehension in pulling in a gigantic linear algebra library for the same reason.
> 
> That said, if we were to implement some kind of matrix transpose, I doubt it would be on some sort of proper matrix type and instead just on an array of SIMD vectors.

This is what I meant indeed. The only thing that fits the scope of a general-purpose SIMD library in my opinion is transposition of `[Simd<T; N>; N]` for all T and N, because doing it right requires knowing which shuffles are implemented in hardware so it's still something that the SIMD portability layer must take care of.

Steps above that, like transposing matrices whose dimensions are not NxN through appropriate masking and scalar special cases, can be taken care of by domain-specific libraries.

**Comment by programmerjake** [comment]
imho transposition, if supported at all, should be supported for all element types for all `NxM` matrices, not just `NxN`.

also, imho transpose should just compile to the llvm intrinsic and if llvm produces non-optimal code that should be fixed in llvm, imho portable-simd should not be a wrapper over `std::arch`.
